server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFastViewFactory.java

```diff
@@ -0,0 +1,201 @@
+            Preconditions.checkNotNull(previews);
+            Preconditions.checkArgument(!previews.isEmpty(), "previews cannot be empty");
+
+            this.previews = previews;
+        }
+
+        @Override
+        public MessageFastView fromMessageResults(Collection<MessageResult> messageResults) throws MailboxException {
+            Helpers.assertOneMessageId(messageResults);
+
+            MessageResult firstMessageResult = messageResults.iterator().next();
+            List<MailboxId> mailboxIds = Helpers.getMailboxIds(messageResults);
+
+            Message mimeMessage = Helpers.parse(firstMessageResult);
+
+            return MessageFastView.builder()
+                .id(firstMessageResult.getMessageId())
+                .mailboxIds(mailboxIds)
+                .blobId(BlobId.of(blobManager.toBlobId(firstMessageResult.getMessageId())))
+                .threadId(firstMessageResult.getMessageId().serialize())
+                .keywords(Helpers.getKeywords(messageResults))
+                .size(firstMessageResult.getSize())
+                .inReplyToMessageId(Helpers.getHeader(mimeMessage, "in-reply-to"))
+                .subject(Strings.nullToEmpty(mimeMessage.getSubject()).trim())
+                .headers(Helpers.toMap(mimeMessage.getHeader().getFields()))
+                .from(Helpers.firstFromMailboxList(mimeMessage.getFrom()))
+                .to(Helpers.fromAddressList(mimeMessage.getTo()))
+                .cc(Helpers.fromAddressList(mimeMessage.getCc()))
+                .bcc(Helpers.fromAddressList(mimeMessage.getBcc()))
+                .replyTo(Helpers.fromAddressList(mimeMessage.getReplyTo()))
+                .date(Helpers.getDateFromHeaderOrInternalDateOtherwise(mimeMessage, firstMessageResult))
+                .preview(previews.get(firstMessageResult.getMessageId()))
+                .build();
+        }
+    }
+
+    private class FromMessageResult implements Helpers.FromMessageResult<MessageFastView> {
+
+        private final MessageFullViewFactory messageFullViewFactory;
+
+        private FromMessageResult(MessageFullViewFactory messageFullViewFactory) {
+            this.messageFullViewFactory = messageFullViewFactory;
+        }
+
+        @Override
+        public MessageFastView fromMessageResults(Collection<MessageResult> messageResults) throws MailboxException {
+            return fromMetaDataWithContent(messageFullViewFactory.toMetaDataWithContent(messageResults));
+        }
+
+        private MessageFastView fromMetaDataWithContent(MessageFullViewFactory.MetaDataWithContent message) throws MailboxException {
+            Message mimeMessage = Helpers.parse(message);
+            MessageContentExtractor.MessageContent messageContent = messageFullViewFactory.extractContent(mimeMessage);
+            Optional<String> mainTextContent = messageFullViewFactory.mainTextContent(messageContent);
+
+            Optional<Preview> preview = mainTextContent.map(Preview::compute);
```

Comment:
Just thought about it but please save the preview (asynchronously? With a `.subscribe()`) so that we don't need to recompute it ever again?

Reply:
I'll do it in a separated PR as it doesn't need this change set to base on

Reply:
#2944